Add special encoding to csv imported struct fields

zcstarr opened the pull request:
…n logic, and changes csv importing to strip csv name fields of invalid struct characters and changes spaces in field names to camel case. i.e. camel case is translated to camelCase.

Additional comments and conversation here. I was struggling to undo my fat finger rebase orz.
https://github.com/attic-labs/noms/pull/2437

aboodman commented:
Nit: we try to follow http://chris.beams.io/posts/git-commit/ for the commit messages.

aboodman commented on go/types/struct.go:
```diff
@@ -169,18 +169,33 @@ func (s1 Struct) Diff(s2 Struct, changes chan<- ValueChanged, closeChan <-chan s
+var escapeRegex = regexp.MustCompile(escapeChar)
+
+var fieldNameComponentRe = regexp.MustCompile("^" + headFieldNamePattern.String() + tailFieldNamePattern.String() + "*")
+var fieldNameRe = regexp.MustCompile(fieldNameComponentRe.String() + "$")
```
Are any of the regexs still used other than `fieldNameRe`? If not, we could just put the whole thing there, simplify.

aboodman replied:
Answering, self: yes, by `EscapeFields`

zcstarr replied:
Yes all of them are in use, headFieldNamePattern and tailFieldNamePattern are used as incremental regex for the encoding logic and fieldNameComponentRe is used by path

aboodman commented on samples/go/csv/read.go:
```diff
@@ -44,6 +46,31 @@ func KindsToStrings(kinds KindSlice) []string {
 
+//EscapeStructFieldFromCSV removes special characters and replaces spaces with camelCasing (camel case turns to camelCase)
+func EscapeStructFieldFromCSV(input string) string {
+
```
Remove blank line

aboodman commented on go/types/struct.go:
```diff
@@ -169,18 +169,33 @@ func (s1 Struct) Diff(s2 Struct, changes chan<- ValueChanged, closeChan <-chan s
+var fieldNameComponentRe = regexp.MustCompile("^" + headFieldNamePattern.String() + tailFieldNamePattern.String() + "*")
+var fieldNameRe = regexp.MustCompile(fieldNameComponentRe.String() + "$")
+
+type encodingFunc func(string, *regexp.Regexp) string
```
I think it is confusing to leave testing the regex to the callee. Why would that not be the responsibility of EscapeFields?

aboodman replied:
Sigh. Answering self again: I see.

aboodman replied:
I see why you are doing this, but I find the signature awkward. As a caller of `EscapeFields`, how do I know what to do with the regex that is passed to `encodingFunc`.

Could you factor this like so instead?

`func QEncodeFieldName(name string) string`

and:

`func CamelCaseFieldName(name string) string`

all the rest of the guts should be package-private implementation details, not exposed to caller.

zcstarr replied:
I think this makes sense and is much cleaner 👍 .

aboodman commented on go/types/struct.go:
```diff
@@ -169,18 +169,33 @@ func (s1 Struct) Diff(s2 Struct, changes chan<- ValueChanged, closeChan <-chan s
+
+type encodingFunc func(string, *regexp.Regexp) string
+
+func EscapeFields(input string, encode encodingFunc) string {
```
Nit: this only escapes one field name, so it should be signular.

aboodman commented on go/types/struct.go:
```diff
@@ -195,13 +210,45 @@ func EscapeStructField(input string) string {
 		return buf.String()
 	}
+	output := EscapeFields(input, encode)
+	verifyFieldName(output)
```
Isn't the output guaranteed to be valid at this point, assuming `EscapeFields` works properly? If so there's no need for this check.

zcstarr replied:
Correct, I suppose if we were to break the encoding function, we'd catch something like that in test in general. Wasn't sure if really bad things happened if we introduced a bug in the encode that was missed in test, but that generated bad data or if we'd have enough checks upstream that we'd just fail to write the struct.

aboodman replied:
We should have thorough unit tests for functions, then callers of those functions can assume they are correct without having to double-check the work.

zcstarr commented:
@aboodman mind giving this another once over

aboodman commented:
Please improve the PR title

aboodman commented:
It should be one sentence, at most 50 chars max

aboodman commented:
see http://chris.beams.io/posts/git-commit/

aboodman commented on go/types/struct.go:
```diff
@@ -195,13 +194,87 @@ func EscapeStructField(input string) string {
+		output += strings.Title(strings.ToLower(escapeFields(field, encode)))
+	}
+
+	if !IsValidStructFieldName(output) {
```
Can you put a little comment here explaining how this can happen? The code is correct, but it's not immediately obvious from reading it why this would occur.

aboodman commented on go/types/struct.go:
```diff
@@ -195,13 +194,87 @@ func EscapeStructField(input string) string {
 	return output
 }
+
+// EscapeStructField escapes names for use as noms structs. Disallowed characters are encoded as
```
Comment out of date.

aboodman replied:
Actually, sorry, comment is correct. But can we just collapse this with QEncodeFieldName? do we need both?

aboodman commented:
Almost there. One more comment/question. BTY.

aboodman commented on go/types/struct.go:
```diff
@@ -215,7 +215,7 @@ func CamelCaseFieldName(input string) string {
 		output += strings.Title(strings.ToLower(escapeFields(field, encode)))
 	}
-
+	//Because we are removing characters, we may generate an invalid field name
```
Maybe include an example? This seems subtle enough that it warrants one.

zcstarr commented:
@aboodman PTAL I changed the camelcase encoding logic a bit to support a few more stripped data edge cases and renamed escapedFields to escapedField.

aboodman commented on go/types/struct.go:
```diff
@@ -180,31 +181,38 @@ var fieldNameRe = regexp.MustCompile(fieldNameComponentRe.String() + "$")
-		output = strings.ToLower(output)
+	//Camelcase field
+	if len(splitField) == 1 {
+		output = splitField[0]
```
Shouldn't `splitField[0]` be `ToLower()`'d?

zcstarr replied:
In this case it assumes that there's only one string and should just preserve the case of the original input. i.e. 

---- Camels ---- should just be Camels vs camels , if that makes sense ?

zcstarr replied:
discussed irl, lowercasing this makes sense especially because is a camelcasing function

aboodman commented on samples/go/csv/read_test.go:
```diff
@@ -148,6 +148,9 @@ func TestEscapeStructFieldFromCSV(t *testing.T) {
 		"💩", "",
 		"11 1💩", "",
+		"-- A B", "aB",
+		"-- A --", "A",
```
Yeah actually I'd expect the result to be "a" here, no?

zcstarr replied:
the thought here is that the dashes are bad, so the algorithm would strip the bad characters and apply camel casing as a best effort. so if we strip the dashes we get "A B" then transforming it to camel case it becomes aB.

zcstarr replied:
uncertain if these expectations are the right one , I can change it if it makes more sense going in a different direction

zcstarr commented:
@aboodman  PTAL

aboodman commented on go/types/struct.go:
```diff
@@ -191,18 +191,18 @@ func CamelCaseFieldName(input string) string {
 	splitField := strings.Fields(strippedField)
-	output := ""
+
+	if len(splitField) == 0 {
```
good catch. i would not have guessed Fields could return empty slice.

aboodman commented:
lgtm
